src/main/java/org/truffleruby/language/methods/Arity.java

```diff
@@ -11,47 +11,52 @@
 
 import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
 
 import org.truffleruby.core.string.StringUtils;
 import org.truffleruby.parser.ArgumentDescriptor;
 import org.truffleruby.parser.ArgumentType;
 
 public class Arity {
 
-    public static final String[] NO_KEYWORDS = StringUtils.EMPTY_STRING_ARRAY;
+    public static final Map<String, Boolean> NO_KEYWORDS = new HashMap<>();
     public static final Arity NO_ARGUMENTS = new Arity(0, 0, false);
     public static final Arity ONE_REQUIRED = new Arity(1, 0, false);
 
     private final int preRequired;
     private final int optional;
     private final boolean hasRest;
     private final int postRequired;
+    private final int optionalKeywordArguments;
     private final boolean hasKeywordsRest;
-    private final String[] keywordArguments;
+    /** Maps a keyword argument to is required property */
+    private final Map<String, Boolean> keywordArguments;
```

Comment:
A HashMap is quite inefficient in terms of memory footprint.
Could we still use a `String[]` and just know whether it's optional or required based on the number of `requiredKeywordArguments` (or the number of `optionalKeywordArguments`)?

On MRI:
```
ruby -e 'def m(a:, b: 3, c:); end; p method(:m).parameters'
[[:keyreq, :a], [:keyreq, :c], [:key, :b]]
```
So required kwargs seem always reported first.

Reply:
If that's too complicated then having a separate `boolean[] requiredKeywords` could be an alternative.

Reply:
> So required kwargs seem always reported first.

Hm, I need to check this.

Anyway, I don't see the problem using `boolean[] \ int requiredIndex` instead of HashMap. Just need to test what better

Reply:
(Also the HashMap would not preserve the order)
`int requiredKeywordArguments` is the best in terms of footprint of course.

Reply:
```sh
$ jt ruby -e 'def m(a:, b: 3, c:); end; p method(:m).parameters
[[:keyreq, :a], [:key, :b], [:keyreq, :c]]
```

The `#parameters` does not use `Arity#getKeywordArguments` so if change only `Arity#keywordArguments`
(prepare ordered by `instanceof RequiredKeywordArgumentValueParseNode` property) it will not affect other methods (like `#paramenters`).

So probably the best option here to have sorted  `ParseNode` in `ArgsParseNode#args`. Or change `ArgsParseNode#getKeywords` to return sorted  keywords. 

Is this what you would like to get?

Reply:
I hacked `ArgsParseNode#getKeywords` to return sorted keywords and added simple loger to see `keywordArguments` in `Arity` object (`System.out.println("Method KWARGS " + Arrays.toString(keywordArguments));`).

Here is the result (before the hacking order of kwargs was the same as in declaration):
```ruby

a1 = proc { |foo:, bar: 1| } 
a2 = proc { |bar: 1, foo: 1| }

def m(a:, b: 3, c:); end;

p "**** START ****"
p method(:m).parameters

# Method KWARGS [bar, foo]
# Method KWARGS [bar, foo]
# Method KWARGS [b, a, c]
# **** START ****
# [[:keyreq, :a], [:key, :b], [:keyreq, :c]]
```

So if we change `ArgsParseNode#getKeywords` it will not affect `#parameters` method.

Reply:
Anyway, I'm not sure it `#parameters` method is what you want to fix in the current PR. If not, we can just add a counter `requiredKeywordArgumentsCount` and instead of saving the index just counting required kwargs in the `for`.

Reply:
I forgot that `parameters` doesn't use `Arity`, good point.
No need to change `#parameters` of course.

`requiredKeywordArgumentsCount` + sorting by required/optional sounds good.
I would have the required kwargs before the optional ones though (seems to match the source order in most cases, also for positional args required typically come before optional)

Reply:
> sorting by required/optional sounds good.

Do you mean `Arity#getKeywordArguments` ? I don't quite understand why we need to sort them. it seems we dont showing these values publicly.

Reply:
Yeah, just counting is enough, no need to sort (we don't know which keyword arguments are optional, but we don't need to know that it seems).